File: src/go/4_BehavioralPatterns/11_Visitor/reflective.go
package visitor

import (
  "fmt"
  "strings"
)

type Expression interface {
  // nothing here!
}

type DoubleExpression struct {
  value float64
}

type AdditionExpression struct {
  left, right Expression
}

func Print(e Expression, sb *strings.Builder) {
  if de, ok := e.(*DoubleExpression); ok {
    sb.WriteString(fmt.Sprintf("%g", de.value))
  } else if ae, ok := e.(*AdditionExpression); ok {
    sb.WriteString("(")
    Print(ae.left, sb)
    sb.WriteString("+")
    Print(ae.right, sb)
    sb.WriteString(")")
  }

  // breaks OCP
  // will work incorrectly on missing case
}

func main() {
  // 1+(2+3)
  e := &AdditionExpression{
    &DoubleExpression{1},
    &AdditionExpression{
      left:  &DoubleExpression{2},
      right: &DoubleExpression{3},
    },
  }
  sb := strings.Builder{}
  Print(e, &sb)
  fmt.Println(sb.String())
}

File: src/js/4_BehavioralPatterns/11_Visitor/reflective.js
class NumberExpression
{
  constructor(value)
  {
    this.value = value;
  }
}

class AdditionExpression
{
  constructor(left, right)
  {
    this.left = left;
    this.right = right;
  }
}

class ExpressionPrinter
{
  print(e, buffer)
  {
    if (e instanceof NumberExpression)
    {
      buffer.push(e.value);
    }
    else if (e instanceof AdditionExpression)
    {
      buffer.push('(');
      this.print(e.left, buffer);
      buffer.push('+');
      this.print(e.right, buffer);
      buffer.push(')');
    }
  }
}

let e = new AdditionExpression(
  new NumberExpression(1),
  new AdditionExpression(
    new NumberExpression(2),
    new NumberExpression(3)
  )
);
let buffer = [];
let ep = new ExpressionPrinter();
ep.print(e, buffer);
console.log(buffer.join(''));